x509/bash.go

```diff
@@ -0,0 +1,50 @@
+//go:build !windows
+
+package x509
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/foundriesio/fioctl/subcommands"
+)
+
+func run(name string, arg ...string) string {
+	cmd := exec.Command(name, arg...)
+	cmd.Stderr = os.Stderr
+	out, err := cmd.Output()
+	subcommands.DieNotNil(err)
+	return string(out)
+}
+
+func CreateFactoryCa(ou string) string {
+	run("./create_ca")
+	return string(readFile(FactoryCaCertFile))
+}
+
+func CreateDeviceCa(cn string, ou string) string {
+	run("./create_device_ca", DeviceCaKeyFile, DeviceCaCertFile)
+	return string(readFile(DeviceCaCertFile))
+}
+
+func SignTlsCsr(csrPemBytes []byte) string {
+	return run("./sign_tls_csr", TlsCsrFile)
+}
+
+func SignCaCsr(csrPemBytes []byte, csrFileName string) string {
+	// If csrFileName contains a "*", csrPemBytes is written to a temp file to be used inside the script
+	var fileName string
+	if strings.Contains(csrFileName, "*") {
+		tmpFile, err := os.CreateTemp("", csrFileName)
+		subcommands.DieNotNil(err)
+		_, err = tmpFile.Write([]byte(csrPemBytes))
+		subcommands.DieNotNil(err)
+		defer tmpFile.Close()
+		fileName = tmpFile.Name()
+		defer os.Remove(fileName)
```

Comment:
From https://go.dev/tour/flowcontrol/13:
```
Deferred function calls are pushed onto a stack. When a function returns, its deferred calls are executed in last-in-first-out order. 
```

So, here, the `os.Remove` will be executed before the `os.Close`; that may lead to _unexpected consequences_.

We actually need to close the file _before_ its name is passed into a script.
(Thus, I see a bug both here and in the original implementation).

So, the below order would be a correct one:
```
... := CreateTemp(...)
defer os.Remove(...)
... = tmpFile.Write(...)
tmpFile.Close()
// After this line you may be sure the file contents were flushed to OS, so we can pass a file name into a script.
```

Reply:
Fixed, thanks.